src/khiva/matrix.cpp

```diff
@@ -746,12 +499,244 @@ void findBestN(af::array profile, af::array index, long m, long n, af::array &di
+    long tsLength = static_cast<long>(qt.dims(0));
+    long nTimeSeries = static_cast<long>(qt.dims(1));
+
+    // Tiling the input data to match the batch size, the time series length and the number of time series
+    af::array a_tiled = af::tile(a, 1, 1, 1, static_cast<unsigned int>(batchSize));
+    af::array sum_q_tiled =
+        af::tile(sum_q, static_cast<unsigned int>(tsLength), static_cast<unsigned int>(nTimeSeries));
+    af::array sum_q2_tiled =
+        af::tile(sum_q2, static_cast<unsigned int>(tsLength), static_cast<unsigned int>(nTimeSeries));
+    af::array mean_t_tiled = af::tile(mean_t, 1, 1, 1, static_cast<unsigned int>(batchSize));
+    af::array sigma_t_tiled = af::tile(sigma_t, 1, 1, 1, static_cast<unsigned int>(batchSize));
+
+    // Required to avoid a division by zero when the standard deviation is zero
+    double eps = (sigma_t_tiled.type() == 0) ? EPSILON * 1e4 : EPSILON;
+    af::array lessThanEpsilon = eps >= sigma_t_tiled;
+    sigma_t_tiled = lessThanEpsilon * lessThanEpsilon.as(sigma_t_tiled.type()) + !lessThanEpsilon * sigma_t_tiled;
+    // Computing the distance
+    af::array dist = a_tiled + (-2 * (qt - sum_q_tiled * mean_t_tiled) / sigma_t_tiled) + sum_q2_tiled;
+    dist = af::sqrt(af::abs(dist));
+
+    // The 1st dimension reflects the number of subsequences of the reference time series.
+    // The 2nd dimension reflects the number of reference time series.
+    // The 3rd dimension reflects the number of query time series
+    // The 4th dimension reflects the batch size of query subsequences from the query time series
+    dist = af::reorder(dist, 3, 1, 0, 2);
+
+    af::min(distance, index, dist, 2);
+}
+
+void mass(af::array q, af::array t, af::array a, af::array mean_t, af::array sigma_t, af::array mask,
+          af::array &distance, af::array &index) {
+    // Normalizing the query sequence. q can contain query sequences from multiple series
+    q = khiva::normalization::znorm(q, EPSILON);
+
+    // Sliding dot product of the subsequence q of all the query time series against all the reference time series
+    // contained in t
+    af::array qt = slidingDotProduct(q, t);
+    // Cumulative sum of all the elements contained in q (for each time series, that is why it is done using the first
+    // dimension)
+    af::array sum_q = af::sum(q, 0);
+    // Cumulative sum of squares of all the elements contained in q (for each time series, that is why it is done using
+    // the first dimension)
+    af::array sum_q2 = af::sum(af::pow(q, 2), 0);
+
+    // Calculate the distance and index profiles for all the combinations of query sequences and reference time series
+    calculateDistanceProfile(qt, a, sum_q, sum_q2, mean_t, sigma_t, mask, distance, index);
+}
+
+void mass(af::array q, af::array t, af::array a, af::array mean_t, af::array sigma_t, af::array &distance,
+          af::array &index) {
+    // Normalizing the query sequence. q can contain query sequences from multiple series
```

Comment:
This code is the same as the above function, couldn't one of the reuse the other one?